processor/src/main/java/org/mapstruct/ap/internal/model/assignment/SetterWrapper.java

```diff
@@ -22,19 +22,41 @@
 import java.util.List;
 
 import org.mapstruct.ap.internal.model.common.Type;
+import org.mapstruct.ap.internal.prism.NullValueCheckStrategyPrism;
+
+import static org.mapstruct.ap.internal.prism.NullValueCheckStrategyPrism.ALWAYS;
 
 /**
- * Wraps the assignment in a target setter.
+ * Wraps the assignment in a target setter. include a null check when
+ *
+ * - Not if source is the parameter iso property, because the null check is than handled by the bean mapping
+ * - Not when source is primitive, you can't null check a primitive
+ * - The source property is fed to a conversion somehow before its assigned to the target
+ * - The user decided to ALLWAYS include a null check
+ * - When there's a source local variable name defined (e.g. nested source properties)
+ * - TODO: The last one I forgot..?
  *
  * @author Sjaak Derksen
  */
 public class SetterWrapper extends AssignmentWrapper {
 
     private final List<Type> thrownTypesToExclude;
+    private final boolean includeNullCheck;
+
+    public SetterWrapper(Assignment rhs, List<Type> thrownTypesToExclude, NullValueCheckStrategyPrism nvms,
+        boolean fieldAssignment, Type targetType) {
+        super( rhs, fieldAssignment );
+        this.thrownTypesToExclude = thrownTypesToExclude;
+        this.includeNullCheck = !rhs.getSourceReference().equals( rhs.getSourceParameterName() )
```

Comment:
I think that the checks should go in some method and maybe somehow add the list items that are there in the class Javadoc.